pkg/store/egressbatchstore/fsstore.go

```diff
@@ -0,0 +1,176 @@
+package egressbatchstore
+
+import (
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"sync"
+
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-car/v2/blockstore"
+	"github.com/multiformats/go-multicodec"
+	"github.com/multiformats/go-multihash"
+	"github.com/storacha/go-libstoracha/capabilities/space/content"
+	"github.com/storacha/go-ucanto/core/receipt"
+	fdm "github.com/storacha/go-ucanto/core/result/failure/datamodel"
+)
+
+const (
+	// DefaultBatchSize is the default maximum size of a receipt batch in bytes.
+	DefaultBatchSize = 100 * 1024 * 1024 // 100MiB
+
+	// currentBatchName is the name of the current batch file.
+	currentBatchName = "egress.car.wip"
+
+	// batchFilePrefix is the prefix for completed batch files.
+	batchFilePrefix = "egress."
+	batchFileSuffix = ".car"
+)
+
+var _ EgressBatchStore = (*fsBatchStore)(nil)
+
+type fsBatchStore struct {
+	mu           sync.Mutex
+	basePath     string
+	curBatchPath string
+	maxBatchSize int64
+}
+
+// NewFSBatchStore creates a new file system based batch receipt store.
+// Batches will be stored in the given basePath.
+// If maxBatchSize is 0, DefaultBatchSize will be used.
+func NewFSBatchStore(basePath string, maxBatchSize int64) (*fsBatchStore, error) {
+	if maxBatchSize <= 0 {
+		maxBatchSize = DefaultBatchSize
+	}
+
+	if err := os.MkdirAll(basePath, 0755); err != nil {
+		return nil, fmt.Errorf("creating egress batch store directory: %w", err)
+	}
+
+	curBatchPath := filepath.Join(basePath, currentBatchName)
+
+	return &fsBatchStore{
+		basePath:     basePath,
+		curBatchPath: curBatchPath,
+		maxBatchSize: maxBatchSize,
+	}, nil
+}
+
+func (s *fsBatchStore) Append(ctx context.Context, rcpt receipt.Receipt[content.RetrieveOk, fdm.FailureModel]) error {
+	if rcpt == nil {
+		return fmt.Errorf("receipt is nil")
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	rwbs, err := blockstore.OpenReadWrite(s.curBatchPath, nil)
```

Comment:
Skimming the car blockstore code leads me to believe we need to call `Close` on this when were done with it iff using CarV2 - should we include that here?